crates/bevy_animation/src/lib.rs

```diff
@@ -479,174 +485,217 @@ impl AnimationPlayer {
 
-    for (idx, part) in parts {
-        let mut found = false;
-        let children = children.get(current_entity).ok()?;
-        if let Some(cached) = path_cache[idx] {
-            if children.contains(&cached) {
-                if let Ok(name) = names.get(cached) {
-                    if name == part {
-                        current_entity = cached;
-                        found = true;
-                    }
-                }
-            }
+    // If an entity was removed from the tree (removed `Parent`), or had its
+    // name removed and was previously an animation target, it's dirty.
+    for entity in entities_with_removed_parents.read().chain(entities_with_removed_names.read()) {
+        if candidate_targets.get(entity).is_ok_and(|(_, _, _, _, target)| target.is_some()) {
+            worklist.push(entity);
         }
-        if !found {
-            for child in children.deref() {
-                if let Ok(name) = names.get(*child) {
-                    if name == part {
-                        // Found a children with the right name, continue to the next part
-                        current_entity = *child;
-                        path_cache[idx] = Some(*child);
-                        found = true;
-                        break;
-                    }
-                }
+    }
+
+    // If an animation player was just added to the tree or changed names, all
+    // its children are dirty.
+    for (_, player, maybe_kids, name) in players.iter() {
+        if player.is_added() || name.is_changed() {
+            if let Some(kids) = maybe_kids {
+                worklist.extend(kids.iter());
             }
         }
-        if !found {
-            warn!("Entity not found for path {:?} on part {:?}", path, part);
-            return None;
-        }
     }
-    Some(current_entity)
-}
 
-/// Verify that there are no ancestors of a given entity that have an [`AnimationPlayer`].
-fn verify_no_ancestor_player(
-    player_parent: Option<&Parent>,
-    parents: &Query<(Has<AnimationPlayer>, Option<&Parent>)>,
-) -> bool {
-    let Some(mut current) = player_parent.map(Parent::get) else {
-        return true;
-    };
-    loop {
-        let Ok((has_player, parent)) = parents.get(current) else {
-            return true;
-        };
-        if has_player {
-            return false;
+    // Iteratively mark all descendants of dirty entities as dirty themselves.
+    // At the end of this process, `dirty_entities` will consist of a forest.
+    let mut dirty_entities: EntityHashSet<Entity> = EntityHashSet::default();
+    while let Some(entity) = worklist.pop() {
+        if dirty_entities.contains(&entity) {
+            continue;
+        }
+        if let Ok((_, _, Some(kids), _, _)) = candidate_targets.get(entity) {
+            worklist.extend(kids.iter().map(|&kid| kid));
         }
+        dirty_entities.insert(entity);
+    }
+
+    // Find the roots of the trees in the forest.
+    worklist = dirty_entities
```

Comment:
Aren't the roots necessarily in the subset `worklist` from above, _before_ we added all children? If so, can't we gather the roots first and add children later? Rather than add all children and then come back to iterate over them and skip them?